tests/test_status.py

```diff
@@ -0,0 +1,454 @@
+)
+@mock.patch.object(status.app.control, "inspect")
+@mock.patch("time.sleep")
+def test_ping_workers_success(
+    mock_sleep, mock_celery_inspect, max_retries, num_failures, sleep_intervals
+):
+    failures = [ConnectionResetError("Connection reset by peer") for _ in range(num_failures)]
+    success = {"celeery@123456": {"ok": "pong"}}
+
+    ping = mock_celery_inspect.return_value.ping
+    ping.side_effect = failures + [success]
+
+    assert status._ping_workers(retries=max_retries) == success
+
+    mock_celery_inspect.assert_called_once()
+    assert ping.call_count == num_failures + 1
+    assert mock_sleep.call_count == len(sleep_intervals)
+    mock_sleep.assert_has_calls(mock.call(t) for t in sleep_intervals)
+
+
+@pytest.mark.parametrize("retries", [0, 1])
+@pytest.mark.parametrize(
+    "ping_result, expect_result",
+    [
+        ({}, []),
+        (
+            {"celery@abcdef": {"ok": "pong"}, "celery@123456": {"ok": "pong"}},
+            # results should be sorted by name
+            [{"name": "celery@123456", "ok": True}, {"name": "celery@abcdef", "ok": True}],
+        ),
+        (
+            {"celery@abcdef": {"ok": "pong"}, "celery@123456": {"error": "sucks at ping-pong"}},
+            [
+                {"name": "celery@123456", "ok": False, "reason": "sucks at ping-pong"},
+                {"name": "celery@abcdef", "ok": True},
+            ],
+        ),
+    ],
+)
+@mock.patch("cachito.web.status._ping_workers")
+def test_workers_status(mock_ping_workers, retries, ping_result, expect_result):
+    mock_ping_workers.return_value = ping_result
+
+    workers = status.workers_status(retries=retries)
+    assert workers == expect_result
+
+    mock_ping_workers.assert_called_once_with(retries=retries)
+
+
+@pytest.mark.parametrize(
+    "worker_ok, failing_services, expect_result",
+    [
+        (False, [], {"gomod": False, "npm": False, "pip": False, "git-submodule": False}),
+        (True, ["rabbitmq"], {"gomod": False, "npm": False, "pip": False, "git-submodule": False}),
+        (True, ["database"], {"gomod": False, "npm": False, "pip": False, "git-submodule": False}),
+        (True, ["athens"], {"gomod": False, "npm": True, "pip": True, "git-submodule": True}),
+        (True, ["nexus"], {"gomod": True, "npm": False, "pip": False, "git-submodule": True}),
+        (
+            True,
+            ["nexus-hoster"],
+            {"gomod": True, "npm": False, "pip": False, "git-submodule": True},
+        ),
+        (
+            True,
+            ["athens", "nexus"],
+            {"gomod": False, "npm": False, "pip": False, "git-submodule": True},
+        ),
+    ],
+)
+def test_can_process(worker_ok, failing_services, expect_result):
+    services = [{"name": service, "ok": False} for service in failing_services]
+    assert status._can_process(TEST_PACKAGE_MANAGERS, services, worker_ok) == expect_result
+
+
+def mock_worker_config(nexus_hoster=False):
```

Comment:
Nitpick: could we move this function to the beginning of the file?

Reply:
Done